### PR TITLE
Add GCCVER define to Toolchain.

### PR DESCRIPTION
**Pull request by cuiziwei1**
## Summary
We hope to find a common place to define GCCVER, so that it only needs to be specified once, which would undoubtedly be the most convenient. https://github.com/apache/nuttx/pull/14130#issuecomment-2408282192.  However, since GCCVER may be used in the Toolchain.defs of each architecture, we have to define GCCVER again.https://github.com/apache/nuttx/pull/14293#issuecomment-2415506922

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



**No, this PR does not meet the NuttX requirements.**

The provided information only includes a build error message and lacks essential details outlined in the NuttX PR requirements. 

**Here's what's missing:**

* **Summary:**  No information is provided about the purpose, functionality, or implementation of the change. 
* **Impact:** The impact on users, the build process, hardware, documentation, security, compatibility, and other considerations is entirely absent.
* **Testing:**  While error logs are partially included, there's no context about the testing environment (host and target), the nature of the tests performed, or the expected behavior.

**To meet the NuttX requirements, the PR needs to provide a comprehensive description of the change, its impact, and thorough testing evidence, following the template provided.**

**Review by pkarashchenko — APPROVED** [review]

**Review by hartmannathan — APPROVED** [review]

File: arch/risc-v/src/cmake/Toolchain.cmake [hunk 1]
```diff
@@ -259,7 +259,16 @@ if(CONFIG_RISCV_TOOLCHAIN STREQUAL GNU_RVG)
...
 
   if(CONFIG_ARCH_RV_ISA_ZICSR_ZIFENCEI)
-    set(ARCHCPUEXTFLAGS ${ARCHCPUEXTFLAGS}_zicsr_zifencei)
+    if(NOT DEFINED GCCVER)
+      execute_process(COMMAND ${CMAKE_CXX_COMPILER} --version
+                      OUTPUT_VARIABLE GCC_VERSION_OUTPUT)
+      string(REGEX MATCH "\\+\\+.* ([0-9]+)\\.[0-9]+" GCC_VERSION_REGEX
+                   "${GCC_VERSION_OUTPUT}")
+      set(GCCVER ${CMAKE_MATCH_1})
+    endif()
```
Comment (anchao) [inline arch/risc-v/src/cmake/Toolchain.cmake:268]: @xuxin930 @xiaoxiang781216 Let's define the get gcc version as a function(cmake, marco in makefile) and remove the duplicate logic from other place

**Review by anchao — CHANGES_REQUESTED** [review]

**Comment by anchao** [comment]
Please make some improvements for revert back commits and make each change more valuable, rather than `revert`,`revert "revert"`,`revert "revert "revert""` in daily work

File: arch/risc-v/src/cmake/Toolchain.cmake (hunk 1 above)
Comment (cuiziwei1) [reply arch/risc-v/src/cmake/Toolchain.cmake:268]: ok

Comment (xiaoxiang781216) [reply arch/risc-v/src/cmake/Toolchain.cmake:268]: let's change ARCH_RV_ISA_ZICSR_ZIFENCEI to n instead:
https://github.com/apache/nuttx/blob/master/arch/risc-v/Kconfig#L414-L427

Comment (xuxin930) [reply arch/risc-v/src/cmake/Toolchain.cmake:268]: > @xuxin930 @xiaoxiang781216 Let's define the get gcc version as a function(cmake, marco in makefile) and remove the duplicate logic from other place

Sure, I'll complete this action.

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/risc-v/src/common/Toolchain.defs [hunk 2]
```diff
@@ -210,7 +210,9 @@ ifeq ($(CONFIG_RISCV_TOOLCHAIN),GNU_RVG)
   endif
 
   ifeq ($(CONFIG_ARCH_RV_ISA_ZICSR_ZIFENCEI),y)
-    ARCHCPUEXTFLAGS := $(ARCHCPUEXTFLAGS)_zicsr_zifencei
+    ifeq ($(shell expr $(GCCVER) \>= 12), 1)
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/Toolchain.defs:213]: So we do not check for `CONFIG_ARCH_TOOLCHAIN_CLANG` here?

Comment (cuiziwei1) [reply arch/risc-v/src/common/Toolchain.defs:213]: It seems that CONFIG_ARCH_TOOLCHAIN_CLANG needs to be added. the latest changes have already added the judgment. please review again，Thanks.

**Comment by xiaoxiang781216** [comment]
please fixtheconflict, @cuiziwei1

**Comment by xiaoxiang781216** [comment]
@cuiziwei1 please fix the conflict.

File: libs/libxx/libcxx.defs [hunk 3]
```diff
@@ -61,6 +61,10 @@ endif
 CXXFLAGS += -Wno-shadow -Wno-sign-compare
 CXXFLAGS += -Wno-attributes -Wno-deprecated-declarations
 
+ifeq ($(GCCVER),)
```
Comment (xiaoxiang781216) [inline libs/libxx/libcxx.defs:64]: why export again

File: libs/libxx/libcxx.cmake [hunk 4]
```diff
@@ -95,6 +95,14 @@ list(APPEND SRCS ${SRCSTMP})
 set(FLAGS -Wno-attributes -Wno-deprecated-declarations -Wno-shadow
           -Wno-sign-compare)
 
+if(NOT GCCVER)
+  execute_process(COMMAND ${CMAKE_C_COMPILER} --version
+                  OUTPUT_VARIABLE GCC_VERSION_INFO)
+  string(REGEX MATCH "[0-9]+\\.[0-9]+" GCC_VERSION ${GCC_VERSION_INFO})
+  string(REGEX REPLACE "\\..*" "" GCCVER ${GCC_VERSION})
+  set(GCCVER ${CMAKE_MATCH_1})
```
Comment (xiaoxiang781216) [inline libs/libxx/libcxx.cmake:103]: ditto

File: arch/arm/src/common/Toolchain.defs [hunk 5]
```diff
@@ -309,11 +309,18 @@ else
    # Wrong warning array subscript [0] is outside array bounds:
    # https://gcc.gnu.org/bugzilla/show_bug.cgi?id=105523
 
+   ifeq ($(GCCVER),)
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/Toolchain.defs:312]: update cmake too

File: arch/arm64/src/Toolchain.defs [hunk 6]
```diff
@@ -180,8 +180,14 @@ ifeq ($(CONFIG_LTO_FULL),y)
   endif
 endif
 
+ifeq ($(GCCVER),)
```
Comment (xiaoxiang781216) [inline arch/arm64/src/Toolchain.defs:183]: why some guard by CONFIG_ARCH_TOOLCHAIN_GNU, but some not?

File: arch/risc-v/src/cmake/Toolchain.cmake [hunk 7]
```diff
@@ -259,7 +259,16 @@ if(CONFIG_RISCV_TOOLCHAIN STREQUAL GNU_RVG)
   endif()
 
   if(CONFIG_ARCH_RV_ISA_ZICSR_ZIFENCEI)
-    set(ARCHCPUEXTFLAGS ${ARCHCPUEXTFLAGS}_zicsr_zifencei)
+    if(NOT DEFINED GCCVER)
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/cmake/Toolchain.cmake:262]: should move outside CONFIG_ARCH_RV_ISA_ZICSR_ZIFENCEI

File: arch/risc-v/src/common/Toolchain.defs [hunk 8]
```diff
@@ -210,7 +210,15 @@ ifeq ($(CONFIG_RISCV_TOOLCHAIN),GNU_RVG)
   endif
 
   ifeq ($(CONFIG_ARCH_RV_ISA_ZICSR_ZIFENCEI),y)
-    ARCHCPUEXTFLAGS := $(ARCHCPUEXTFLAGS)_zicsr_zifencei
+    ifeq ($(GCCVER),)
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/Toolchain.defs:213]: ditto

File: arch/xtensa/src/lx6/Toolchain.defs [hunk 9]
```diff
@@ -111,6 +111,10 @@ ifneq ($(CONFIG_LIBCXXTOOLCHAIN),y)
   ARCHCXXFLAGS += -nostdinc++
 endif
 
+ifeq ($(CONFIG_ARCH_TOOLCHAIN_GNU), y)
```
Comment (xiaoxiang781216) [inline arch/xtensa/src/lx6/Toolchain.defs:114]:
```suggestion
ifeq ($(CONFIG_ARCH_TOOLCHAIN_GNU),y)
```

File: arch/xtensa/src/lx7/Toolchain.defs [hunk 10]
```diff
@@ -115,6 +115,10 @@ ifneq ($(CONFIG_LIBCXXTOOLCHAIN),y)
   ARCHCXXFLAGS += -nostdinc++
 endif
 
+ifeq ($(CONFIG_ARCH_TOOLCHAIN_GNU), y)
```
Comment (xiaoxiang781216) [inline arch/xtensa/src/lx7/Toolchain.defs:118]:
```suggestion
ifeq ($(CONFIG_ARCH_TOOLCHAIN_GNU),y)
```

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**